src/Peachpie.CodeAnalysis/Semantics/Graph/SingleBlockWalker.cs

```diff
@@ -0,0 +1,561 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+using Pchp.CodeAnalysis.Semantics.TypeRef;
+
+namespace Pchp.CodeAnalysis.Semantics.Graph
+{
+    /// <summary>
+    /// Visitor used to traverse a single block and its adjacent edge.
+    /// </summary>
+    public abstract class SingleBlockWalker<T> : GraphVisitor<T>
```

Comment:
another visitor?

Reply:
It was needed to properly analyse each block separately (including the condition in its adjacent edge), without traversing the others, enabling to separate the order of block traversal from the visitor itself. It was inserted in the hierarchy between `GraphWalker` and `GraphVisitor`, dividing `GraphWalker`'s functionality a bit.

Reply:
can just `GraphVisitor` traverse like that? or we need two different visitor semantics

Reply:
We can make `GraphWalker` traverse like that (`GraphVisitor`'s `Visit*` methods are empty). Due to its lack of infinite recursion prevention it cannot be used in a standalone way anyway. We can move its traversal of the edges to `GraphExplorer` (`AnalysisWalker` overrides all the edge visiting anyway). In that case, we might better rename it to something like `BlockWalker`?

Reply:
sounds better than fourth visitor class

Reply:
Removed in 93776b70682f6c429231968456235de00574b5a6, after basing the two new analyses on `AnalysisWalker`.